File: packages/python/catalogmx/catalogs/sat/cfdi_4/metodo_pago.py
"""Catálogo c_MetodoPago"""

from catalogmx.catalogs.sat.cfdi_4._resolver_views import code_description_rows


class MetodoPagoCatalog:
    """Catálogo de Métodos de Pago del SAT (c_MetodoPago)"""

    _data: list[dict] | None = None
    _by_code: dict[str, dict] | None = None

    @classmethod
    def _load_data(cls) -> None:
        """Carga lazy desde el artefacto canónico CFDI 4.0."""
        if cls._data is None:
            cls._data = code_description_rows("cfdi_40_metodos_pago")
            cls._by_code = {item["code"]: item for item in cls._data}

    @classmethod
    def get_metodo(cls, code: str) -> dict | None:
        """Obtiene un método de pago por su código"""
        cls._load_data()
        return cls._by_code.get(code.upper())

    @classmethod
    def is_valid(cls, code: str) -> bool:
        """Valida si un código de método de pago es válido"""
        return cls.get_metodo(code) is not None

    @classmethod
    def get_all(cls) -> list[dict]:
        """Obtiene todos los métodos de pago"""
        cls._load_data()
        return cls._data.copy()

File: packages/dart/lib/src/catalogs/sat/cfdi/metodo_pago.dart
/// Catálogo de Métodos de Pago (SAT)
///
/// Catálogo c_MetodoPago del SAT para CFDI 4.0
library;

import 'package:catalogmx/src/catalogs/base_catalog.dart';

/// Método de Pago Catalog
class MetodoPagoCatalog {
  static List<Map<String, dynamic>>? _data;
  static Map<String, Map<String, dynamic>>? _byClave;

  static void _loadData() {
    if (_data != null) return;

    final jsonData = BaseCatalog.loadJsonDataSync('sat/cfdi_4.0/metodo_pago.json');

    // Handle both list and dict formats
    if (jsonData.isNotEmpty && jsonData.first.containsKey('metodos_pago')) {
      _data = (jsonData.first['metodos_pago'] as List)
          .map((e) => e as Map<String, dynamic>)
          .toList();
    } else {
      _data = jsonData;
    }

    // Build index
    _byClave = {
      for (var item in _data!)
        item['clave'] as String: item,
    };
  }

  /// Obtiene todos los métodos de pago
  static List<Map<String, dynamic>> getAll() {
    _loadData();
    return List.from(_data!);
  }

  /// Obtiene método de pago por clave
  static Map<String, dynamic>? getByClave(String clave) {
    _loadData();
    return _byClave![clave];
  }

  /// Valida si una clave de método de pago existe
  static bool isValid(String clave) {
    return getByClave(clave) != null;
  }
}
